ssl/statem/statem_clnt.c

```diff
@@ -1136,6 +1138,24 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
                session_id_len);
     }
 
+    /* Session version and negotiated protocol version should match
+     * (This may not be the case when reusing old session)
+     */
+    if (s->version != s->session->ssl_version) {
```

Comment:
Nope, `if` above will set it to `1`. We are kind of accounting for a faulty server here, that will accept session and negotiate higher protocol version than supported by the session.